File: Meraki_practice/Rest.scala
import scala.collection.mutable.ListBuffer;
import scala.collection.mutable.Map;
import org.apache.http.impl.client.HttpAClientBuilder;

object Rest {
	def mapToJson(map: Map[String, String]): String = {
		var result = "{";
		map.keys.foreach(k => {
			if (result != "") { result = s"${result},"; }
			result = s"\"${k}\":\"${map(k)}\"";
		});
		result += "}";
		return result;
	}

	def invokeGet(url: String): String = {
		val httpClient = httpClientBuilder.create().build();
		val httpResponse = httpClient.execute(new HttpGet(url));
		val entity = httpResponse.getEntity();
		var content = "";
		if (entity != null) {
	  		val inputStream = entity.getContent();
	  		content = io.Source.fromInputStream(inputStream).getLines.mkString;
	  		inputStream.close;
		}
		httpClient.getConnectionManager().shutdown();
		return content;
  	}

	def invokePost(url: String, body: Map[String, String]): String = {
		val httpClient = httpClientBuilder.create().build();
		val bodyJson = mapToJson(body);
		val httpResponse = httpClient.execute(new HttpPOst(url, bodyJson));
		val entity = httpResponse.getEntity();
		var content = "";
		if (entity != null) {
			val inputStream = entity.getContent();
			content = io.Source.fromInputStream(inputStream).getLines.mkString;
			inputStream.close;
		}
		httpClient.getConnectionManager().shutdown();
		return content;
	}

	def main (args: Array[String]) = {
		var methodType: String				= "GET";
		var endpoint  : String				= "";
		var body	  : Map[String, String] = Map();
		var headers   : Map[String, String] = Map();

		for (i <- 0 to args.size - 1) {
			if (args(i) == "-m") {
				methodType = args(i + 1);
			} 
			else if (args(i) == "--endpoint") {
				endpoint = args(i + 1);
			} 
			else if (args(i) == "-h") {
				var pairs = args(i + 1).split(",");
				for (pair <- pairs) {
					var comps = pair.split(":");
					if (comps.size > 1) {
						headers(comps(0)) = comps(1);
					}
				}
			} 
			else if (args(i) == "-d") {
				var pairs = args(i + 1).split(",");
				for (pair <- pairs) {
					var comps = pair.split(":");
					if (comps.size > 1) {
						body(comps(0)) = comps(1);
					}
				}
			}
		}

		var rsp = "";
		methodType match {
			case "GET" => {
				rsp = invokeGet(endpoint);
			}
			case "POST" => {
				rsp = invokePost(endpoint, body);
			}
			case whoa => {
				println(s"Method (${methodType}) not valid...");
				System.exit(-1);
			}
		}
		println(s"Response: ${rsp}");	
	}
}

File: Meraki_practice/rest.rb
#!/usr/bin/ruby
###############################################################################
# Name        : rest.rb                                                       #
# Author      : Abel Gancsos                                                  #
# Version     : v. 1.0.0.0                                                    #
# Description : Practice working with REST.                                   #
###############################################################################
require "uri";
require "net/http";

class Rester
	@methodType=nil;@endpoint=nil;@headers=nil;@body=nil;
	
	def initialize(params={})
		@methodType = params["-m"] != nil ? params["-m"] : "GET";
		@endpoint = params["--endpoint"] != nil ? params["--endpoint"] : "https://api.nasa.gov/planetary/apod?api_key=DEMO_KEY";
		@headers = {};
		@body = {};
		if (params["-h"] != nil)
			pairs = params["-h"].split(",");
			for i in 0..pairs.length
				comps = pairs[i].split(":");
				if (comps.length == 2)
					@headers[comps[0]] = comps[1];
				end
			end
		end
		if (params["-d"] != nil)
			pairs = params["-d"].split(",");
			for i in 0..pairs.length
				comps = pairs[i].split(":");
				if (comps.length == 2)
					@body[comps[0]] = comps[1];
				end
			end
		end
	end

	def invoke_get(endpoint, headers)
		req = Net::HTTP::Get.new(URI(@endpoint));
		for k in @headers.keys
			req[k] = @headers[k];
		end
		rsp = Net::HTTP.start(req.uri.host, :use_ssl => req.uri.scheme == 'https') do |http|
			http.request(req)
		end
		return rsp;
	end

	def invoke_post(endpoint, headers, body)
		req = Net::HTTP::Post.new(URI(@endpoint));
		req.body = "#{body}";
		for k in @headers.keys
			req[k] = @headers[k];
		end
		rsp = Net::HTTP.start(req.uri.host, :use_ssl => req.uri.scheme == 'https') do |http|
			http.request(req)
		end
		return rsp;
	end

	def invoke()
		rsp = nil;
		case (@methodType)
			when "GET"
				rsp = self.invoke_get(@endpoint, @headers);
			when "POST"
				rsp = self.invoke_post(@endpoint, @headers, @body);
			else
				print "Method #{@methodType} not valid...\n";
				exit(-1);
		end
		print "Response: #{rsp.body}";
	end 
end

params = {};
for i in 0..ARGV.length - 1
	params[ARGV[i]] = ARGV[i + 1];
end

session = Rester.new(params);
session.invoke();
